internal/persistence/sql/persister.go

```diff
@@ -28,121 +29,101 @@ import (
+	//namespaceMigrations embed.FS
 
 	_ persistence.Persister = &Persister{}
 )
 
-func NewPersister(dsn string, l *logrusx.Logger, namespaces namespace.Manager, tracer *tracing.Tracer) (*Persister, error) {
-	pop.SetLogger(l.PopLogger)
-
-	p := &Persister{
-		namespaces: namespaces,
-		l:          l,
-		dsn:        dsn,
-		tracer:     tracer,
-	}
+func NewPersister(reg dependencies, nid uuid.UUID) (*Persister, error) {
+	//pop.SetLogger(reg.Logger().PopLogger)
 
-	var err error
-	p.conn, err = p.newConnection(nil)
+	pop.Debug = true
+	conn, err := reg.PopConnection()
 	if err != nil {
 		return nil, err
 	}
 
-	return p, nil
-}
-
-func (p *Persister) newConnection(options map[string]string) (c *pop.Connection, err error) {
-	var opts []instrumentedsql.Opt
-	if p.tracer.IsLoaded() {
-		opts = []instrumentedsql.Opt{
-			instrumentedsql.WithTracer(opentracing.NewTracer(true)),
-			instrumentedsql.WithOmitArgs(),
-		}
+	namespaceIDCache, err := ristretto.NewCache(&ristretto.Config{
+		// TODO probably make this configurable
+		NumCounters: 10 * 1000,
+		MaxCost:     1000,
+		BufferItems: 64,
+	})
+	if err != nil {
+		return nil, errors.WithStack(err)
 	}
-	pool, idlePool, connMaxLifetime, connMaxIdleTime, cleanedDSN := sqlcon.ParseConnectionOptions(p.l, p.dsn)
-	connDetails := &pop.ConnectionDetails{
-		URL:                       sqlcon.FinalizeDSN(p.l, cleanedDSN),
-		IdlePool:                  idlePool,
-		ConnMaxLifetime:           connMaxLifetime,
-		ConnMaxIdleTime:           connMaxIdleTime,
-		Pool:                      pool,
-		Options:                   options,
-		UseInstrumentedDriver:     p.tracer != nil && p.tracer.IsLoaded(),
-		InstrumentedDriverOptions: opts,
+
+	p := &Persister{
+		d:                reg,
+		nid:              nid,
+		conn:             conn,
+		namespaceIDCache: namespaceIDCache,
 	}
 
-	bc := backoff.NewExponentialBackOff()
-	bc.MaxElapsedTime = time.Minute * 5
-	bc.Reset()
+	return p, nil
+}
 
-	if err := backoff.Retry(func() (err error) {
-		c, err = pop.NewConnection(connDetails)
-		if err != nil {
-			p.l.WithError(err).Error("Unable to connect to database, retrying.")
-			return errors.WithStack(err)
-		}
+func NewMigrationBox(c *pop.Connection, logger *logrusx.Logger, tracer *tracing.Tracer) (*popx.MigrationBox, error) {
+	return popx.NewMigrationBox(fsx.Merge(migrations, networkx.Migrations), popx.NewMigrator(c, logger, tracer, 0))
+}
 
-		if err := c.Open(); err != nil {
-			p.l.WithError(err).Error("Unable to open the database connection, retrying.")
-			return errors.WithStack(err)
-		}
+func (p *Persister) Connection(ctx context.Context) *pop.Connection {
+	return popx.GetConnection(ctx, p.conn.WithContext(ctx))
+}
 
-		if err := c.Store.(interface{ Ping() error }).Ping(); err != nil {
-			p.l.WithError(err).Error("Unable to ping the database connection, retrying.")
-			return errors.WithStack(err)
-		}
+func (p *Persister) CreateWithNetwork(ctx context.Context, v interface{}) error {
```

Comment:
Wow this diff is super messy... Please especially look into this function here, and whether it is a good approach. IMO this is a valid use-case for `reflect`, but we should think about that.

Reply:
So generally this looks pretty smart. Personally I feel that reflection should not belong into strict enforcement code of things that do isolation. Since we have panics here, it kind of feels ok though!